Client/Frontend/Home/ActivityStreamTopSitesCell.swift

```diff
@@ -0,0 +1,393 @@
+    A modified version of http://stackoverflow.com/a/34167915
+ */
+
+class HorizontalFlowLayout: UICollectionViewLayout {
+    var itemSize = CGSize.zero
+    private var cellCount = 0
+    private var boundsSize = CGSize.zero
+    private var insets = UIEdgeInsetsZero
+    private let minimumInsets: CGFloat = 20
+    var numberOfPages = 0
+
+    override func prepareLayout() {
+        cellCount = self.collectionView!.numberOfItemsInSection(0)
+        boundsSize = self.collectionView!.bounds.size
+    }
+
+    override func collectionViewContentSize() -> CGSize {
+        return collectionViewSizeForRect(boundsSize)
+    }
+
+    func maxVerticalItemsCount() -> Int {
+        let verticalItemsCount =  Int(floor(boundsSize.height / (itemSize.height + insets.top)))
+        if let delegate = self.collectionView?.delegate as? ASHorizontalScrollDelegate {
+            return delegate.numberOfVerticalItems(verticalItemsCount)
+        } else {
+            return verticalItemsCount
+        }
+    }
+
+    func maxHorizontalItemsCount() -> Int {
+        let horizontalItemsCount =  Int(floor(boundsSize.width / (itemSize.width + insets.left)))
+        if let delegate = self.collectionView?.delegate as? ASHorizontalScrollDelegate {
+            return delegate.numberOfHorizontalItems(horizontalItemsCount)
+        } else {
+            return horizontalItemsCount
+        }
+    }
+    func collectionViewSizeForRect(contentSize: CGSize) -> CGSize {
+        let horizontalItemsCount = maxHorizontalItemsCount()
+        let verticalItemsCount = maxVerticalItemsCount()
+
+
+        // Take the number of cells and subtract its space in the view from the height. The left over space is the white space.
+        // The left over space is then devided evenly into (n + 1) parts to figure out how much space should be inbetween a cell
+        var verticalInsets = (contentSize.height - (CGFloat(verticalItemsCount) * itemSize.height)) / CGFloat(verticalItemsCount + 1)
+        var horizontalInsets = (contentSize.width - (CGFloat(horizontalItemsCount) * itemSize.width)) / CGFloat(horizontalItemsCount + 1)
+
+        // We want a minimum inset to make things not look crowded. We also don't want uneven spacing.
+        // If we dont have this. Set a minimum inset and recalculate the size of a cell
+        if horizontalInsets < minimumInsets || horizontalInsets != verticalInsets {
+            verticalInsets = minimumInsets
+            horizontalInsets = minimumInsets
+            itemSize.width = (contentSize.width - (CGFloat(horizontalItemsCount + 1) * horizontalInsets)) / CGFloat(horizontalItemsCount)
+            itemSize.height = itemSize.width
+        }
+
+        let itemsPerPage = verticalItemsCount * horizontalItemsCount
+
+        numberOfPages = Int(ceil(Double(cellCount) / Double(itemsPerPage)))
+
+
+        insets = UIEdgeInsets(top: verticalInsets, left: horizontalInsets, bottom: verticalInsets, right: horizontalInsets)
```

Comment:
Nit: Maybe just personal taste but how come there are double newlines throughout? Do we really need two lines of emptiness?